src/filter.cpp

```diff
@@ -226,3 +171,63 @@ SEXP dplyr_mask_eval_all_filter(SEXP quos, SEXP env_private, SEXP s_n, SEXP env_
 
   return keep;
 }
+
+SEXP new_logical(int n, int value) {
+  SEXP x = PROTECT(Rf_allocVector(LGLSXP, n));
+  int* p_x = LOGICAL(x);
+  for (int i = 0; i < n; i++) {
+    p_x[i] = value;
+  }
+  UNPROTECT(1);
+  return x;
+}
+
+SEXP dplyr_reduce_lgl_or(SEXP df, SEXP n_) {
+  int n = INTEGER(n_)[0];
+  R_xlen_t ncols = XLENGTH(df);
+
+  if (ncols == 1) {
+    return VECTOR_ELT(df, 0);
+  }
+
+  SEXP reduced = PROTECT(new_logical(n, FALSE));
+  int* p_reduced = LOGICAL(reduced);
+  if (ncols > 0) {
+    const SEXP* p_df = VECTOR_PTR_RO(df);
+    for (R_xlen_t j = 0; j < ncols; j++) {
+      int* p_df_j = LOGICAL(p_df[j]);
+      for (int i = 0; i < n; i++) {
+        p_reduced[i] = p_reduced[i] == TRUE || p_df_j[i] == TRUE;
```

Comment:
```suggestion
        p_reduced[i] = p_reduced[i] || p_df_j[i];
```

Can the comparison be removed?

Reply:
I think not because of NA_LOGICAL but I have to check

Reply:
if I make these two changes, I get: 

``` r
library(dplyr, warn.conflicts = FALSE)

df <- expand.grid(
  x = c(TRUE, FALSE, NA), y = c(TRUE, FALSE, NA)
)
filter(df, x, y)
#>      x    y
#> 1 TRUE TRUE
filter(df, x & y)
#>      x    y
#> 1 TRUE TRUE
filter(df, if_all(c(x,y), identity))
#>      x    y
#> 1 TRUE TRUE
#> 2   NA TRUE
#> 3 TRUE   NA
#> 4   NA   NA

filter(df, x | y)
#>       x     y
#> 1  TRUE  TRUE
#> 2 FALSE  TRUE
#> 3    NA  TRUE
#> 4  TRUE FALSE
#> 5  TRUE    NA
filter(df, if_any(c(x,y), identity))
#>       x     y
#> 1  TRUE  TRUE
#> 2 FALSE  TRUE
#> 3    NA  TRUE
#> 4  TRUE FALSE
#> 5    NA FALSE
#> 6  TRUE    NA
#> 7 FALSE    NA
#> 8    NA    NA
```

<sup>Created on 2021-01-28 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

otherwise, I get: 

``` r
library(dplyr, warn.conflicts = FALSE)

df <- expand.grid(
  x = c(TRUE, FALSE, NA), y = c(TRUE, FALSE, NA)
)
filter(df, x, y)
#>      x    y
#> 1 TRUE TRUE
filter(df, x & y)
#>      x    y
#> 1 TRUE TRUE
filter(df, if_all(c(x,y), identity))
#>      x    y
#> 1 TRUE TRUE

filter(df, x | y)
#>       x     y
#> 1  TRUE  TRUE
#> 2 FALSE  TRUE
#> 3    NA  TRUE
#> 4  TRUE FALSE
#> 5  TRUE    NA
filter(df, if_any(c(x,y), identity))
#>       x     y
#> 1  TRUE  TRUE
#> 2 FALSE  TRUE
#> 3    NA  TRUE
#> 4  TRUE FALSE
#> 5  TRUE    NA
```

<sup>Created on 2021-01-28 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>